### ISSUE TITLE
Efficiency: Counting pixels via CPP: Iterate over pixels once and attribute to each range, instead of iterating over range + pixels

### ISSUE THREAD

**Issue by Claudius-Appel** [Claudius-Appel/duflor#16]
In the cpp- pixel count function setup, currently:

- for each hsv-set, loop over the entire range of pixels and return all hits.
- this means that all x\*y pixels are checked for all N hsv ranges.

**Instead**:
- would it be faster to iterate over all N pixels and for each pixel check if it fulfills conditions of each range. In the case of a hit, it would be attributed to that ranges xy-records.
- in theory, this would mean iterating over all pixels only 1 time, instead of N times. 

*Prereqs*:
- confirm that the cpp solution does not record xy-coords
    - however if it does, maybe I can record x/y pairs in a dataframe, and name the x/y with a suffix of the range it belongs to.

*Issues*:
- how to handle `area_complete`? One could _assume_ it to be the sum of `area_green`+`area_drought`, but that is not inherently true. If it were, we could skip it and just merge the returns of `area_green` and  `area_drought`. 

_Originally posted by @Claudius-Appel in https://github.com/Claudius-Appel/duflor/issues/9#issuecomment-2017415276_

**Comment by Claudius-Appel** [comment]
First results suggest that this is indeed faster, circumstantially.

```R
# don't touch these.
times <- 1
foreach_verbose <- F
external_gc <- F
cluster_type <- NA

# modify these as instructed.
# Assuming a 16 core-CPU, I suggest initially testing setups of 4/8/12/14 cores.
used_cores <- 3
```
```R
> print(t.1  - t.0)
Time difference of 7.037594 mins

> cat("run on",used_cores,"cores")
run on 3 cores
> df1
                             Time in NS  N Time in S
GC(), bundled PA,par       138901431100 32 138.90143
GC(), no bundled PA,par     76594404700 32  76.59440
no GC(), bundled PA,par    123583343100 32 123.58334
no GC(), no bundled PA,par  80737812100 32  80.73781
> parrallel
                             Time in NS  N Time in S estimated par-time for 200 images
GC(), bundled PA,par       138901431100 32 138.90143                          868.1339
GC(), no bundled PA,par     76594404700 32  76.59440                          478.7150
no GC(), bundled PA,par    123583343100 32 123.58334                          772.3959
no GC(), no bundled PA,par  80737812100 32  80.73781                          504.6113
```

**Comment by Claudius-Appel** [comment]
implemented in 9cd88c4c4f546a2b423d299c5aff7f8b58606b36
